src/panels/lovelace/cards/energy/hui-energy-sankey-card.ts

```diff
@@ -0,0 +1,432 @@
+          types.battery.map((source) => source.stat_energy_to)
+        ) || 0;
+
+      nodes.push({
+        id: "battery_in",
+        label: this.hass.localize(
+          "ui.panel.lovelace.cards.energy.energy_distribution.battery"
+        ),
+        value: totalBatteryIn,
+        tooltip: `${formatNumber(totalBatteryIn, this.hass.locale)} kWh`,
+        color: "var(--energy-battery-in-color)",
+        index: 1,
+      });
+      nodes.forEach((node) => {
+        // Link all sources to battery_in
+        if (node.index === 0) {
+          links.push({
+            source: node.id,
+            target: "battery_in",
+          });
+        }
+      });
+
+      homeNode.value -= totalBatteryIn;
+    }
+
+    // Add grid return if available
+    if (types.grid && types.grid[0].flow_to) {
+      const totalToGrid =
+        calculateStatisticsSumGrowth(
+          this._data.stats,
+          types.grid[0].flow_to.map((flow) => flow.stat_energy_to)
+        ) ?? 0;
+
+      nodes.push({
+        id: "grid_return",
+        label: this.hass.localize(
+          "ui.panel.lovelace.cards.energy.energy_distribution.grid"
+        ),
+        value: totalToGrid,
+        tooltip: `${formatNumber(totalToGrid, this.hass.locale)} kWh`,
+        color: "var(--energy-grid-return-color)",
+        index: 1,
+      });
+      nodes.forEach((node) => {
+        // Link all sources to grid_return
+        if (node.index === 0) {
+          links.push({
+            source: node.id,
+            target: "grid_return",
+          });
+        }
+      });
+
+      homeNode.value -= totalToGrid;
+    }
+
+    // Group devices by areas and floors
+    const areas: Record<string, { value: number; devices: Node[] }> = {
+      no_area: {
+        value: 0,
+        devices: [],
+      },
+    };
+    const floors: Record<string, { value: number; areas: string[] }> = {
+      no_floor: {
+        value: 0,
+        areas: ["no_area"],
+      },
+    };
+    let untrackedConsumption = homeNode.value;
+    const computedStyle = getComputedStyle(this);
+    prefs.device_consumption.forEach((device, idx) => {
+      const entity = this.hass.entities[device.stat_consumption];
+      const value =
+        device.stat_consumption in this._data!.stats
+          ? calculateStatisticSumGrowth(
+              this._data!.stats[device.stat_consumption]
+            ) || 0
+          : 0;
+      if (value <= 0) {
+        return;
+      }
+      untrackedConsumption -= value;
+      const deviceNode: Node = {
+        id: device.stat_consumption,
+        label:
+          device.name ||
+          getStatisticLabel(
+            this.hass,
+            device.stat_consumption,
+            this._data!.statsMetadata[device.stat_consumption]
+          ),
+        value,
+        tooltip: `${formatNumber(value, this.hass.locale)} kWh`,
+        color: getGraphColorByIndex(idx, computedStyle),
+        index: 4,
+      };
+
+      const deviceArea = entity?.area_id;
```

Comment:
The device area can indeed be in the entity registry, but if that is not set it should be looked up in the device registry